certbot/lock.py

```diff
@@ -27,34 +34,86 @@ def lock_dir(dir_path):
-    This lock file is released when the locked file is closed or the
-    process exits. It cannot be used to provide synchronization between
-    threads. It is based on the lock_file package by Martin Horcicka.
-
+    """
+    Platform independent file lock system.
+    LockFile accepts a parameter, the path to a file acting as a lock. Once the LockFile,
+    instance is created, the associated file is 'locked from the point of view of the OS,
+    meaning that if another instance of Certbot try at the same time to acquire the same lock,
+    it will raise an Exception. Calling release method will release the lock, and make it
+    available to every other instance. Upon exit, current Certbot release also all the locks.
+    This allows to protect a file or a directory to be concurrently accessed and modified
+    by two Certbot instances in parallel.
+    LockFile is platform independent: it will proceed to the appropriate OS lock mechanism
+    depending on Linux or Windows.
     """
     def __init__(self, path):
-        """Initialize and acquire the lock file.
+        # type: (str) -> None
+        """
+        Create a LockFile instance on the given file path, and acquire lock.
+        :param str path: the path to the file that will hold a lock
+        """
+        self._path = path
+        mechanism = _UnixLockMechanism if POSIX_MODE else _WindowsLockMechanism
+        self._lock_mechanism = mechanism(path)
 
-        :param str path: path to the file to lock
+        self.acquire()
 
+    def __repr__(self):
+        # type: () -> str
+        repr_str = '{0}({1}) <'.format(self.__class__.__name__, self._path)
+        if self.is_locked():
+            repr_str += 'acquired>'
+        else:
+            repr_str += 'released>'
+        return repr_str
+
+    def acquire(self):
+        # type: () -> None
+        """
+        Acquire the lock on the file, forbidding any other Certbot instance to acquire it.
         :raises errors.LockError: if unable to acquire the lock
+        """
+        self._lock_mechanism.acquire()  # type: ignore
 
+    def release(self):
+        # type: () -> None
         """
-        super(LockFile, self).__init__()
-        self._path = path
-        self._fd = None
+        Release the lock on the file, allowing any other Certbot instance to acquire it.
+        """
+        if self.is_locked():
+            self._lock_mechanism.release()  # type: ignore
 
-        self.acquire()
+    def is_locked(self):
+        # type: () -> bool
+        """
+        Check if the file is currently locked.
+        :return: True if the file is locked, False otherwise
+        """
+        return self._lock_mechanism.is_locked()  # type: ignore
 
-    def acquire(self):
-        """Acquire the lock file.
 
-        :raises errors.LockError: if lock is already held
-        :raises OSError: if unable to open or stat the lock file
+# Unix locking mechanism
+# ~~~~~~~~~~~~~~~~~~~~~~
+class _UnixLockMechanism(object):
```

Comment:
I personally think most design patterns have their place including inheritance.

To be clear though, I wasn't proposing a major change here. I was just noticing that code like `__init__` and `is_locked` is identical in both locking mechanisms. It's not much, but if we added these methods to `_BaseLockMechanism` which you added, it would result in less code.

Do you still have concerns about this? I'm personally not too worried about whether we do this or not.

Reply:
Yes I just wanted to avoid things like NotImplementedException, abc, and be able to have an agnostic object exposed to the rest of Certbot code without making weird things.

To be more precise also about classes, it is things like abc, or various implementation of interfaces contract, that make me feels a little confortable.

Apart that, yes, I still need an internal base class, I will factor the code that can be there.